nexus/src/app/sagas/vpc_subnet_create.rs

```diff
@@ -0,0 +1,605 @@
+
+        match &result {
+            Some(Ok(_)) => break,
+            // Allow NUM_RETRIES retries, after the first attempt.
+            //
+            // Note that we only catch IPv6 overlaps. The client
+            // always specifies the IPv4 range, so we fail the
+            // request if that overlaps with an existing range.
+            Some(Err(InsertVpcSubnetError::OverlappingIpRange(IpNet::V6(
+                _,
+            )))) if retryable => debug!(
+                log,
+                "autogenerated random IPv6 range overlap";
+                "subnet_id" => ?subnet_id,
+                "ipv6_block" => %ipv6_block
+            ),
+            _ => {}
+        }
+    }
+
+    (match result {
+        None => Err(external::Error::internal_error(
+            "Attempted to create VPC subnet without any IPv6 allocation",
+        )),
+        Some(Err(InsertVpcSubnetError::OverlappingIpRange(IpNet::V6(_))))
+            if retryable =>
+        {
+            // TODO-monitoring TODO-debugging
+            //
+            // We should maintain a counter for this occurrence, and
+            // export that via `oximeter`, so that we can see these
+            // failures through the timeseries database. The main
+            // goal here is for us to notice that this is happening
+            // before it becomes a major issue for customers.
+            error!(
+                log,
+                "failed to generate unique random IPv6 address \
+                range in {} retries",
+                num_retries;
+                "vpc_id" => ?vpc_id,
+                "subnet_id" => ?subnet_id,
+            );
+            Err(external::Error::internal_error(
+                "Unable to allocate unique IPv6 address range \
+                for VPC Subnet",
+            ))
+        }
+        // Overlapping IPv4/explicit v6 range, which is a client error.
+        Some(Err(e @ InsertVpcSubnetError::OverlappingIpRange(_))) => {
+            Err(e.into_external())
+        }
+        Some(Err(InsertVpcSubnetError::External(e))) => Err(e),
+        Some(Ok(v)) => Ok(v),
+    })
+    .map_err(ActionError::action_failed)
+}
+
+async fn svsc_create_subnet_undo(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+    let opctx = crate::context::op_context_for_saga_action(
+        &sagactx,
+        &params.serialized_authn,
+    );
+
+    let (authz_subnet, db_subnet) =
+        sagactx.lookup::<(authz::VpcSubnet, db::model::VpcSubnet)>("subnet")?;
+
+    let res = osagactx
+        .datastore()
+        .vpc_delete_subnet_raw(&opctx, &db_subnet, &authz_subnet)
+        .await;
+
+    match res {
+        Ok(_) | Err(external::Error::ObjectNotFound { .. }) => {
+            let _ = osagactx
+                .datastore()
+                .vpc_increment_rpw_version(&opctx, params.authz_vpc.id())
+                .await;
```

Comment:
The RPW version bump here is needed to account for the fact that *some* nexus could have seen incomplete state and pushed a route out. A new version is required to invalidate that.